### ISSUE TITLE
Add print speed setting for overhang

### ISSUE THREAD

**Issue by devzero** [alexrj/Slic3r#328]
I'm having trouble with overhangs curling up in my prints and I've noticed that slowing down the print can sometimes reduce this curling.  Could you detect overhanging like you do for bridging and offer an additional print speed configuration setting for the times that you are overhanging?

Thanks,
-Chris

**Comment by VanessaE** [comment]
I don't mean to necro-post (does that even count when it's on Github?) but this feature is much-needed, especially if you're printing PLA at higher temperatures.

**Comment by clearbucketLabs** [comment]
almost 2017, this sort of feature is like a basic 101 setting for printing with plastic plastic.     i guess i can take a stab at it but it would be faster by someone that knows the code base.

**Comment by lordofhyphens** [comment]
Well, looks like most of the bits are already there to detect overhangs and the flow is already broken out, but PerimeterGenerator() needs a new constructor to submit an overhang flow to it as well and everywhere the constructor is used probably needs to use the new constructor. 

Flow.hpp needs a new role for overhang flow if you want to break it out completely, followed by a new config option for it (I have notes in the wiki for adding new config options), as well as some sane default in the .cpp. You can probably crib from bridges to get it set up. 

Also would need a flow ratio for overhangs (width doesn't make sense if there's nothing beneath it). 

For speed, you need to set it in Gcode.cpp, function extrude. The current path's flow role is exposed, near as I can tell, as path.first().role (it's an enumeration). You will need to figure out what to do when you have an overhang and a small perimeter at the same time.  The small perimeter speed is set around line 411. 

You can try passing --overhangs on Slic3r invocation to use bridge flows for overhangs for right now just to see if this is really what you want. 

@clearbucketLabs Congrats, you're "hired". The rest of us here are volunteers in our spare time, so things get done when they get done, and I'd be much more interested in reviewing code from someone else and helping them get up to speed and contributing than just doing everything myself.

**Comment by bubnikv** [comment]
You may check the EdgeGrid class in the prusa3d fork and how it is used to
detect overhangs for the seam placement. The EdgeGrid class sorts the
contours into a grid structure and then it enables fast calculation of a
signed distance to the contour, so the speed could then be controlled
proportionally to the overhang distance.

**Comment by alexrj** [comment]
There's a bit of confusion overhere, and the topic is not well defined.

First, I ask people who request this feature to agree on a **definition of "overhang"** in this context. Right now Slic3r has a feature which detects portions of perimeters where the nozzle is fully outside the lower layer contour. For those portions, bridge flow and bridge speed (as well as fan) are applied. That feature is called *Detect bridging perimeters* (`--overhangs` from command line).
Isn't that feature enough?

If it's not, please tell what's your definition of overhang, and what exactly should happen beyond using a dedicated speed. Some of you are talking about flow, but beware that some time ago Slic3r used to use the bridge flow also for much smaller overhangs, and the result was horrible. Also note that changes of speed just for some portions of a perimeter result in a pressure lag, so they are not ideal.

Let's first define the problem before talking about code...

**Comment by lordofhyphens** [comment]
What I think is desired here is the ability to monkey around with speed
and/or fan settings for areas of the print that were unsupported a few
layers below--basically to be able to hack one's way into a WorksForMe
"solution".

I agree about the problem needing better definition.

**Comment by bananapunk** [comment]
Actually what most of us refer to as overhang is partially unsupported perimeter. For example upside down pyramid is pretty much all overhang. What would be great is to have ability to control speed and flow in those areas only. At the moment, to print overhangs perfectly, we need to slow down the whole print to 20 mm/s. Especially overhang corners are a problem, yoda ears for example. unless you print slow , pla tends to curl up, so at the best youll get ugly perimeters, and worst print will get knocked off. Here's quite a good thread of ultimaker users trying to figure it out 
https://ultimaker.com/en/community/5094-raised-edges

And a video illustrating the problem 
https://m.youtube.com/watch?v=jXYxpWr5tGo

**Comment by clearbucketLabs** [comment]
Pretty much bang on. With pla you can blast the fan at everything and it "can" work but it's mixed results and does not apply to every plastic, especially ABS and a few others.  anything that shrinks can benefit.   i don't think there could ever be too many settings to tweak when laying plastic.  instead of brute force it should be based on feature detection of the model,  overhangs, holes, etc....  a bunch of problems can just be solved in slicer (some in FW and HW but..) that makes printing at least slightly more reliable and quicker.   

solutions to these overhangs/curling are all currently "WorksForMe" solutions like rotating the model, moving it to a different spot on the bed, dropping extrusion temp for the whole print or entire layer, adding support under so it cools and sticks without curling etc....

**Comment by bubnikv** [comment]
It would really be interesting to find out, why the slow down helps:

1) Is it that the newly laid threads of plastics are stretched slowly, so, the polymer has more time to reorient its molecules?

2) Is it that the hot end has more time to iron down the already curled layers below?

3) Is it just that the fan has more time to cool down the curling perimeters, when the print is slowed down?

My impression is, you want to slow down over overhangs because of inadequate cooling fan. I also think, that spending more time ironing the overhangs with the hot end is counter productive as it re-heats the overhangs. But I may be wrong and the hypothesis 1) is correct.

Someone to do experiments?

**Comment by bananapunk** [comment]
Its not the inadequate cooling, i tried some crazy setups with portable aircon blowing ice cold air onto external fan etc. More cooling does help somewhat, but by brute force- forcing plastic to freeze before it gets pulled. I have a feeling that too much plastic gets deposited due to no flow control in marlin. That plastic is also only partially supported from underneath, so because its got not much to stick to, it sticks to hot end. Going slow, 20 mms or so, means theres consistent flow, so that helps. Also cooling rapidly prevents extra plastic from following the hotend, if it gets frozen fast enough. Now, since we cant predict the flow that well in slicer, and marlin's flow control is screwed up and turned off, it would be great if we could have the slicer detect overhangs, and especially sharp turns on overhangs, and slow down an X distance before.

**Comment by bubnikv** [comment]
@bananapunk What do you mean with "no flow control in marlin"? I actually hacked Marlin at Prusa3D, so I know the firmware inside out. Do you mean the "pressure advance"?

> it would be great if we could have the slicer detect overhangs, and especially sharp turns on overhangs, and slow down an X distance before.

Yes, I agree with that. There is a "small perimeter speed" settings, but it only triggers for small perimeters, not small "ears" sticking out of large perimeters.

**Comment by bananapunk** [comment]
Thats right, pressure advance. Last time i checked the alghoritm was incomplete and the libraries never actually get called.

**Comment by tcm0116** [comment]
I've been messing around with printing spheres lately, and I've discovered that in addition to cooling, slowing down the extruder can make a significant difference in the quality of the overhang.

The upside-down pyramid @bananapunk referred to is overhang with a constant angle, whereas the lower hemisphere of a sphere is all overhang ranging from 0 degrees at the bottom to 90 degrees at the equator. In this case, I've found that more cooling and slower speeds are required for the bottom of the sphere than at the equator. As such, my recommendation would be to add the ability to control cooling and speed as a function of the angle of the overhang such that lower angles (essentially bridges) would have max cooling and use very low speeds, higher angles (vertical perimeters) would have min cooling and faster speeds, and the two would be modified linearly as the angle changes.

In my current experiment, the bottom of the sphere is clipped, giving it a base to sit on, and I'm using a modifier mesh to slow down the perimeters to 15 mm/s for the bottom portion where the overhangs are the most extreme. That's my solution when printing 6 concurrently, but for printing one, I just configured the cooling parameters such that the minimum layer time as 60 seconds and the minimum speed was 10 mm/s, which yielded almost perfect results. However, I think that my recommendation above would provide a much more efficient way or printing the sphere (or any overhang) as it would allow the speed and cooling to be much more dynamic.

This is a view of one of my earlier attempts that did not adjust the speed at the bottom of the sphere. While not a great picture, you can see the roughness of the front section.

![screenshot_20170320-150911](https://cloud.githubusercontent.com/assets/1868773/24119555/a6e08a64-0d7f-11e7-9873-d1fef65d0462.png)

This is a view of a set of 6 printing, and the bottoms of the spheres are significantly improved by slowing the perimeters down to 15 mm/s using a modifier mesh,

![img_20170311_123029](https://cloud.githubusercontent.com/assets/1868773/24119267/9436fa16-0d7e-11e7-9b6e-06594144ab6e.jpg)

Here's a view of a completed item. This one was printed by itself, so the perimeters were printed closer to 10 mm/s at the bottom (at 60 seconds per layer).

![img_20170310_172612](https://cloud.githubusercontent.com/assets/1868773/24119272/970f946e-0d7e-11e7-9ce6-92a6e3e23cbd.jpg)

**Comment by VanessaE** [comment]
Overhangs need to be detected reliably, and there are five settings that need to be applied to them:
* print speed
* acceleration (to help make sharp corners)
* flow ratio (similar to bridging)
* fan speed
* how far away from the overhang to start and stop applying the fan and speed settings (perhaps a gradual change in speed is warranted)

Some kind of threshold setting would be needed to determine just how much of an overhang needs to happen for the above settings to go into effect (similar to the threshold used to determine if supports are needed).

As referenced in another issue, this model http://www.thingiverse.com/thing:58218 is useful for evaluating these sorts of things.

**Comment by curiouspl2** [comment]
+1 ... just slowing down helps significantly. perhaps because there is just plainly less pull force when pulling slower (plastic stretches slowly) or because of extra filament coming out of nozzle due to ooze , compensating for pull. 

whatever the reason, if slowing down help, why not doing it?

**Comment by bananapunk** [comment]
you can set it up for slow down in simple objects like this example, however when printing anything organic, overhangs come in and out all over the place, setting up zones won't help. So the slicer should analyze the model and set the speed or extrusion multiplier accordingly, slowing down before overhang starts, in order to even out the change of pressure.

**Comment by lordofhyphens** [comment]
@bananapunk it's probably feasible to tag overhang using code similar to the support code.

**Comment by bananapunk** [comment]
that's right! only the changes have to anticipate overhang, otherwise due to rapid slowdown the flow will be too much.

**Comment by lordofhyphens** [comment]
If someone wants to muck around with the current support detection code they can take a look, but the true solution would need to live in the c++ side of things, so that detection would need to exist in libslic3r.

Probably would need to define an ExtrusionRole for overhangs and figure out how to reference part of a polyline (as we don't want to break loops because that would add seams).

**Comment by simonlee** [comment]
I've tripled the speed of my printer by way of many customizations to the hardware and firmware while keeping the quality on par or better. The biggest issue are overhangs where the printer must be slowed down otherwise the curls happen and leads to a distorted result or even a failure. Having to slow down entire layers pretty much negates the speed advantages. If slic3r could detect overhangs of specific angle thresholds then apply a speed to it, that would be amazing. I would then be able to print max speeds for all features without compromising the quality of the print or the overall time spent on the print. I agree that this is an important feature to have as it should greatly reduce printing time while maintaining high quality prints.

**Comment by pavel-ruban** [comment]
Common guys, 8 years passed, this is a really useful feature please implement it.

Condition to detect overhang:
currently printed perimeter is external/internal one and nozzle is NOT fully outside BUT *partially* outside  the lower layer contour. 

It's very different to bridging & should use another set of settings to allow 2 different set of controls for bridging & overhangs for at least these settings:
* print speed
* fan
not mandatory but would be beneficial to have as well:
* nozzle temperature
* feed rate



example of print that requires it to avoid curling where you can see partially overlapped perimeter contours for external or internal perimeters, example has 3 layers  where |__| characters set corresponds to single nozzle line:

    |__|
  |__|
|__|

Or

|__|
  |__|
    |__|


Also it would be beneficial to have some multiplier depending on the overhand %, e.g. if overhand is 70% (70% is beyond/outside the below perimeter contour) then (if checkbox clicked) use this multiplier  % of overhand * CUSTOM_MULTIPLIER_FLOAT * PARAMETER (Fan / Speed etc considering inverting for slow down speed or increasing fan rate etc)

@alranel
